### PR TITLE
Check if user has authenticated themselves thru email

### PR DESCRIPTION
**Pull request by harishan-r**
added a check that verifies that the user is authenticated before allowing them onto home page. User must go into their email now and click the verification link before being allowed onto the app. 

## Notion ticket link

[User Verification](https://www.notion.so/uwblueprintexecs/User-Verification-f4620f59ea49484d9f342d788bcce82b?pvs=4)

## Implementation description
* login and sign up pages calls backend api to check if user has verified themselves thru email. 

## Steps to test
1. Do a test where user signs up and attempts to log in without verifying themselves
2. then do a test where they actually log in

## What should reviewers focus on?
* 

## Checklist
- [ ] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ ] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: frontend/src/components/auth/Signup.tsx
Comment (98ZhaoJeffrey) [inline frontend/src/components/auth/Signup.tsx:1]: There is a bug here where if I signup, refresh the page, it would redirect me to an authenticated path.

File: frontend/src/components/auth/Login.tsx
Comment (98ZhaoJeffrey) [inline frontend/src/components/auth/Login.tsx:1]: Same issue as the one in signup, refresh allows me to login

Comment (98ZhaoJeffrey) [reply frontend/src/components/auth/Login.tsx:1]: Ok the error you are having here can be resolved if you pull from main since I have made those changes from my branch with the role query. Basically you are still setting the user in the authAPIClient login and sign up. 

tldr: pull from main, your code is good :D

File: frontend/src/components/auth/Signup.tsx
Comment (98ZhaoJeffrey) [reply frontend/src/components/auth/Signup.tsx:1]: read above comment in login.tsx

**Review by 98ZhaoJeffrey — COMMENTED** [review]
pull from main to solve the refresh bug

**Comment by carolynzhang18** [comment]
+1 what Jeffrey noticed about how reloading the page allows all users to enter regardless of verification status 🕵️

Here's what (I'm pretty sure) is going on:
- `AuthAPIClient.login`: When a user logs in, their info is added to local storage (peep Chrome dev tools > Application > Local Storage > http://localhost:3000/). An unverified user is still a legitimate user according to Firebase, so the `login` function doesn't hit an error and the `localStorage.setItem` runs.
- `App.tsx`: When we reload the site, we are re-rendering `App.tsx`. In this component, we use the user stored in local storage (if any) to initialize `authenticatedUser`, which is then given to the `AuthContext.Provider`.

Putting these together, what's happening is `AuthAPIClient.login` adds the unverified user to local storage (because their account _is_ valid!), and then reloading the page causes `App.tsx` to initialize the `AuthContext` using the local storage value. At that point, as far as our app knows, it has an authenticated user. Your `isUserAuth` check in `Login.tsx` successfully blocks the `authenticatedUser` update momentarily, but as soon as we hit reload, the `AuthContext` is reinitialized by pulling the local storage object... Annoying :/

The solution seems to be to sync up the local storage setting part with the `setAuthenticatedUser` part. As Jeffrey mentioned, he made some changes (now in main) that accomplish this. +1 you should run a `git pull origin main` 🕺

There's one catch though - your `/isUserVerified/:email` route requires a bearer token to be sent via the HTTP headers (that's what `getAccessToken` returns). You'll see in your `AuthAPIClient.isUserAuth` function, you're currently pulling that from local storage... which obviously won't work once we make the local storage setting part conditional 😅 To work around this chicken and egg situation, I would suggest adding another parameter `accessToken` to `isUserAuth` that can be pass in instead of the local storage object. When you call on this function from `Login.tsx`, you can pass in `user.accessToken`.

Sorry for the yapping, TLDR: pull from main + pass the user's access token to make the API call 👍

**Review by achandrabalan — APPROVED** [review]
Looks Good !

**Review by 98ZhaoJeffrey — APPROVED** [review]

**Review by carolynzhang18 — COMMENTED** [review]
some nits + run `yarn run fix` :  )

File: frontend/src/components/auth/Login.tsx [hunk 1]
```diff
@@ -52,8 +52,12 @@ const Login = (): React.ReactElement => {
...
       return;
     }
-    localStorage.setItem(AUTHENTICATED_USER_KEY, JSON.stringify(user));
-    setAuthenticatedUser(user);
+  
+    const isUserAuth = await authAPIClient.isUserAuth(email, user.accessToken);
+    if (isUserAuth) {
+      localStorage.setItem(AUTHENTICATED_USER_KEY, JSON.stringify(user));
+      setAuthenticatedUser(user);
+    }
```
[the remark this answers is not in the document]
Comment (harishan-r) [reply frontend/src/components/auth/Login.tsx]: I did this instead: 

`const isUserAuth = user?.accessToken ? await authAPIClient.isUserAuth(email, user.accessToken) : null;`
Have to perform a check if user.access token is present

**Review by carolynzhang18 — APPROVED** [review]
image
